base/reflection.jl

```diff
@@ -233,6 +234,55 @@ a concrete type that can have instances, or the empty type `Union{}`.
 isleaftype(t::ANY) = (@_pure_meta; (isa(t, DataType) && t.isleaftype) || t === Union{})
 
 """
+    Base.isabstract(T)
+
+Determine whether `T` was declared as an abstract type (i.e. using the
+`abstract` keyword).
+"""
+function isabstract(t::ANY)
+   @_pure_meta
+   t = unwrap_unionall(t)
+   isa(t,DataType) && t.abstract
+end
+
+"""
+    Base.parameter_upper_bound(t::UnionAll, idx)
+
+Determine the upper bound of a type parameter in the underlying type. E.g.:
+```jldoctest
+julia> immutable Foo{T<:AbstractFloat, N}
+           x::Tuple{T, N}
+       end
+
+julia> Base.parameter_upper_bound(Foo, 1)
+AbstractFloat
+
+julia> Base.parameter_upper_bound(Foo, 2)
+Any
+```
+"""
+function parameter_upper_bound(t::UnionAll, idx)
+   @_pure_meta
+   rewrap_unionall(unwrap_unionall(t).parameters[idx], t)
+end
+
+"""
+    Base.unparameterized_type(t)
+
+Given a (partially) specialized type, return the unspecialized type, e.g.:
+```jldoctest
+julia> immutable Foo{T}
+           x::T
+       end
+
+julia> Base.unparameterized_type(Foo{Float64})
+Foo
+```
+"""
+unparameterized_type(t::DataType) = t.name.wrapper
+unparameterized_type(t::UnionAll) = unparameterized_type(unwrap_unionall(t))
```

Comment:
maybe we could try to give this a better name, although I'm not sure what. I think this is essentially `supertype`? So maybe something like `basetype`? In general, I think we've tried to avoid providing this, since it makes it tempting to assume you could re-parameterize it (even though `basetype(T){tparam}` is not generally a well-defined operation)